### PR TITLE
Android: using AccessibilityNodeInfo#addAction to announce Expandable/Collapsible State

### PR DESCRIPTION
**Pull request by fabriziobertoglio1987**
## Summary

>Expandable and Collapsible are unique in the Android Accessibility API, in that they are not represented as properties on the View or AccessibilityNodeInfo, but are only represented as AccessibilityActions on the AccessibilityNodeInfo. This means that Talkback determines whether or not a node is "expandable" or "collapsible", or potentially even both, by looking at the list of AccessibilityActions attached to the AccessibilityNodeInfo.

>When setting the accessibilityState's expandable property, it should correlate to adding an action of either AccessibilityNodeInfoCompat.ACTION_EXPAND or AccessibilityNodeInfoCompat.ACTION_COLLAPSE on the AccessibilityNodeInfo. This work should be done in the ReactAccessibilityDelegate class's

>Currently, this feature is being "faked" by appending to the contentDescription in the BaseViewManager class. This should be removed when this feature is implemented properly.

fixes https://github.com/facebook/react-native/issues/30841

## Changelog

[Android] [Fixed] - using AccessibilityNodeInfo#addAction to announce Expandable/Collapsible State

## Test Plan

- On some components, the state expanded/collapsed is properly announced on focus, on some it is not.
- On some components only the expanded/collapsed state is announced, and not other component text.
- Upon change, state change is not always announced.
- The accessibilityState's "expanded" field does not seem to work on all element types (for example, it has no effect on 's).
- using accessibilityActions it is possible to add an action for expand/collapse, but these are treated as custom actions and must have their own label defined, rather than using Androids built in expand/collapse actions, which Talkback has predefined labels for.

https://snack.expo.io/0YOQfXFBi

Tests  15th August 2022:
- Paper [Tests](https://github.com/facebook/react-native/pull/34353#issuecomment-1217425302)
- Fabric [Tests](https://github.com/facebook/react-native/pull/34353#issuecomment-1217781734)

Tests 6th September 2022:
- [Button which keeps control of extended/collapsed state in JavaScript with onAccessibilityAction, accessibilityActions and accessibiltyState (Paper)](https://github.com/facebook/react-native/pull/34353#issuecomment-1237601847) 
- [TouchableWithoutFeedback keeps control of extended/collapsed state in Android Widget (Paper)](https://github.com/facebook/react-native/pull/34353#issuecomment-1237616304)
- [TouchableWithoutFeedback keeps control of extended/collapsed state in Android Widget (Fabric)](https://github.com/facebook/react-native/pull/34353#issuecomment-1237624755)
- [TouchableOpacity announces visible text and triggers expanded/collapsed with onPress and accessiblity menu (Fabric)](https://github.com/facebook/react-native/pull/34353#issuecomment-1237627857)

Announcing state with custom actions on Fabric (FAIL). 
The issue is not a regression from this PR, as documented in https://github.com/facebook/react-native/pull/34353#issuecomment-1207744977. It will be fixed in a separate PR.

### CODE REVIEW

**Comment by fabriziobertoglio1987** [comment]
[] - review impact on https://github.com/facebook/react-native/pull/33690

**Comment by fabriziobertoglio1987** [comment]
<details><summary>using accessibilityActions it is possible to add an action for expand/collapse, but these are treated as custom actions and must have their own label defined, rather than using Android built-in expand/collapse actions, which Talkback has predefined labels for.</summary>
<p>

https://user-images.githubusercontent.com/24992535/183358727-aaa3d1b8-c872-4ed0-8a45-906ce3850d53.mp4

</p>
</details>

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/BaseViewManager.java [hunk 1]
```diff
@@ -338,12 +338,6 @@ private void updateViewContentDescription(@NonNull T view) {
             && value.asBoolean()) {
           contentDescription.add(view.getContext().getString(R.string.state_busy_description));
         } else if (state.equals(STATE_EXPANDED) && value.getType() == ReadableType.Boolean) {
-          contentDescription.add(
```
Comment (fabriziobertoglio1987) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/BaseViewManager.java:341]: removes the functionality on Paper

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactAccessibilityDelegate.java [hunk 2]
```diff
@@ -414,6 +416,8 @@ private static void setState(
               context.getString(
                   boolValue ? R.string.state_on_description : R.string.state_off_description));
         }
+      } else if (state.equals("expanded") && value.getType() == ReadableType.Boolean) {
+        info.addAction(value.asBoolean() ? AccessibilityNodeInfoCompat.ACTION_EXPAND : AccessibilityNodeInfoCompat.ACTION_COLLAPSE);
```
Comment (fabriziobertoglio1987) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactAccessibilityDelegate.java:420]: this does not work on paper renderer

**Comment by fabriziobertoglio1987** [comment]
<details><summary>PR Branch - using accessibilityActions it is possible to add an action for expand/collapse, but these are treated as custom actions and must have their own label defined, rather than using Android built-in expand/collapse actions, which Talkback has predefined labels for.</summary>
<p>

https://user-images.githubusercontent.com/24992535/183358727-aaa3d1b8-c872-4ed0-8a45-906ce3850d53.mp4

https://user-images.githubusercontent.com/24992535/183359244-9ff11682-a625-48f5-810b-d198aa173cce.mp4

solved with https://github.com/facebook/react-native/pull/34353/commits/d58a2dbe88ee63bc55dd6c908770c98000b573d4

Screen Shot 2022-08-08 at 21 37 59

</p>
</details>

**Comment by fabriziobertoglio1987** [comment]
<details><summary>extended/collapsed is announced on state change on Paper</summary>
<p>

https://user-images.githubusercontent.com/24992535/185028855-17791b9d-a586-420e-9c10-29673ba224e2.mp4

</p>
</details>

**Comment by fabriziobertoglio1987** [comment]
<details><summary>extended/collapsed is announced on state change on Fabric (PASS)</summary>
<p>


https://user-images.githubusercontent.com/24992535/185082749-0cb2ad4d-59de-4338-a22c-b3563d0a98b8.mp4


</p>
</details>

<details><summary>announce the visible text on Paper (FAIL)</summary>
<p>

https://user-images.githubusercontent.com/24992535/185088735-4d559be8-7183-444f-b140-7ebc262128ee.mp4

</p>
</details>

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactAccessibilityDelegate.java [hunk 3]
```diff
@@ -414,6 +414,8 @@ private static void setState(
...
```
Comment (blavalla) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactAccessibilityDelegate.java:418]: Maybe I'm misreading this, but isn't this adding an action of "ACTION_EXPAND" when the state is "expanded"? 

It should do the opposite, when a state is "expanded" the only available action should be "ACTION_COLLAPSE" and when a state is "collapsed" the only available action should be "ACTION_EXPAND".

**Review by blavalla — CHANGES_REQUESTED** [review]

**Comment by fabriziobertoglio1987** [comment]
<details><summary>Button which keeps control of extended state in JavaScript with onAccessibilityAction, accessibilityActions and accessibiltyState (Paper)</summary>
<p>

https://user-images.githubusercontent.com/24992535/188535897-b24609ec-d134-4987-ba83-138984403e0d.mp4

</p>
</details>

**Comment by fabriziobertoglio1987** [comment]
<details><summary>TouchableWithoutFeedback keeps control of extended/collapsed state in Android Widget (Paper)</summary>
<p>



https://user-images.githubusercontent.com/24992535/188540174-5bc19b96-da72-43de-ad7c-e9dbebd1b68b.mp4


</p>
</details>

**Comment by fabriziobertoglio1987** [comment]
<details><summary>TouchableWithoutFeedback keeps control of extended/collapsed state in Android Widget (Fabric)</summary>
<p>

https://user-images.githubusercontent.com/24992535/188542275-9ea6e531-0a84-48a3-a2a5-f1d8aa90d033.mp4

</p>
</details>

**Comment by fabriziobertoglio1987** [comment]
<details><summary>TouchableOpacity announces visible text and triggers expanded/collapsed with onPress and accessiblity menu (Fabric)</summary>
<p>

https://user-images.githubusercontent.com/24992535/188542974-83e23b4c-6c99-4594-b30c-622bc8a36170.mp4

</p>
</details>

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactAccessibilityDelegate.java (hunk 3 above)
Comment (fabriziobertoglio1987) [reply ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactAccessibilityDelegate.java]: Thanks a lot. I added the suggested changes. I also added these changes:
- updated the AccessibilityExample ([videos](https://github.com/facebook/react-native/pull/34353#issuecomment-1237601847))
- added `setTag` and `getTag` to retrieve and update the `accessibility_state_expanded` in ReactAccessibilityDelegate

https://github.com/facebook/react-native/blob/b80bcb3104c903f97974d90db3827b997af9c7cc/ReactAndroid/src/main/java/com/facebook/react/uimanager/BaseViewManager.java#L268-L276

https://github.com/facebook/react-native/blob/b80bcb3104c903f97974d90db3827b997af9c7cc/ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactAccessibilityDelegate.java#L228-L235

- added the logic in performAccessibilityAction to announce the change in expanded collapsed state when using accessibility menu without implementation of [onAccessibilityAction](https://reactnative.dev/docs/next/view#onaccessibilityaction).

https://github.com/facebook/react-native/blob/b80bcb3104c903f97974d90db3827b997af9c7cc/ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactAccessibilityDelegate.java#L365-L371

- updated the summary with the new video test cases
- further investigated the issue with onAccessibilityAction not triggered on Fabric (summary available [here](https://github.com/reactwg/react-native-new-architecture/discussions/74#discussioncomment-3566068)).

**Review by blavalla — APPROVED** [review]
This looks great! Thanks for being so thorough here.

**Pull request closed without merging**